### PR TITLE
RD-1237 Remove duplicated query string code

### PR DESCRIPTION
**Pull request by leon-good-life**
Jira ticket: https://cloudifysource.atlassian.net/browse/RD-1237

## Description

**Development Notes: (Updated)**

The implementation of `getUrlWithQueryString` in `backend/sharedUtils.ts` is the only implementation of creating a url with query params across the frontend, backend and system tests.

In `/backend` there is only one function that creates a url with query params: `getUrlWithQueryString` in`backend/sharedUtils.ts` and it is called in multiple places, including in frontend.

In `/app` and `/widgets` there are 2 functions:
-  `appendQueryParam` function inside `app/utils/shared/UrlUtils.ts` that is called in multiple places. `getUrlWithQueryString` is used inside.
- `buildActualUrl` method in `app/utils/External.ts`, `getUrlWithQueryString` is used inside..

In `test/cypress/support/resource_commons.ts` there is a function `appendQueryParam`. I attempted to reuse `appendQueryParam` from `app/utils/shared/UrlUtils.ts` but it caused system test fails.

## Screenshots / Videos
N/A

## Checklist

- [x] My code follows the [UI Code Style](https://cloudifysource.atlassian.net/l/c/x8fq902p).
- [x] I verified that all tests and checks have passed.
- [x] I verified if that change requires any update in the [documentation](https://cloudifysource.atlassian.net/l/c/4XKtPocy).
- [x] I added proper labels to this PR.

## Tests
second attempt - all system tests passed:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3422/pipeline

first attempt:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3421/pipeline

 

## Documentation
I'm not sure if I need to add something to: https://docs.cloudify.co/latest/developer/writing_widgets/widget-apis/

### CODE REVIEW

File: app/utils/External.ts [hunk 1]
```diff
@@ -300,10 +300,7 @@ export default class External {
 
     // eslint-disable-next-line class-methods-use-this
     protected buildActualUrl(url: string, data?: Record<string, any>) {
-        const queryString = data
-            ? (url.indexOf('?') > 0 ? '&' : '?') + stringifyQueryString(data, { sort: false })
-            : '';
-        return `${url}${queryString}`;
+        return UrlUtils.appendQueryParam(url, data);
```
Comment (leon-good-life) [inline app/utils/External.ts:303]: The reason that I didn't remove the `buildActualUrl` from `app\utils\External.ts` is that it is overridden in `app/utils/Internal` and `app/utils/WidgetBackend.ts` classes which extends `app\utils\External.ts`:
```
   buildActualUrl(path, data) {
       return super.buildActualUrl(`/wb/${path}`, data);
   }
```

File: app/utils/shared/UrlUtils.ts [hunk 2]
```diff
@@ -1,9 +1,15 @@
...
 export default class UrlUtils {
-    static appendQueryParam(url: string, param: string, value: string | number) {
-        return `${url}${url.indexOf('?') > 0 ? '&' : '?'}${param}=${value}`;
+    static appendQueryParam(url: string, data?: Record<string, any>) {
+        const prefix = url.includes('?') ? '&' : '?';
+        let queryString = '';
+        if (data) {
+            queryString = prefix + stringify(data, { sort: false });
+        }
+        return `${url}${queryString}`;
```
Comment (kubama) [inline app/utils/shared/UrlUtils.ts:12]: A bit simpler:
```suggestion
        if (!data) return url;
        
        const prefix = url.includes('?') ? '&' : '?';
        const queryString = prefix + stringify(data, { sort: false });
        return `${url}${queryString}`;
```

File: test/cypress/support/resource_commons.ts [hunk 3]
```diff
@@ -11,7 +11,8 @@ declare global {
 }
 
 function appendQueryParam(url: string, param: string, value: string) {
-    return `${url}${url.indexOf('?') > 0 ? '&' : '?'}${param}=${value}`;
+    const prefix = url.includes('?') ? '&' : '?';
+    return `${url}${prefix}${param}=${encodeURIComponent(value)}`;
 }
```
Comment (kubama) [inline test/cypress/support/resource_commons.ts:16]: Why not reuse the common implementation here? I guess it was the goal of the ticket :slightly_smiling_face:.

File: backend/handler/services/common.ts [hunk 4]
```diff
@@ -3,10 +3,11 @@ import { stringify } from 'query-string';
 import type { QueryStringParams } from '../../types';
 
 export function getUrlWithQueryString(url: string, params: QueryStringParams = {}) {
+    const prefix = url?.includes('?') ? '&' : '?';
     let result = url;
 
     if (!isEmpty(params)) {
-        result += (url.indexOf('?') > 0 ? '&' : '?') + stringify(params, { sort: false });
+        result += prefix + stringify(params, { sort: false });
     }
```
Comment (kubama) [inline backend/handler/services/common.ts:11]: This way we still have separate implementation for frontend and backend. I would place the common function in `sharedUtils.ts` and reuse it on both layers.

File: app/utils/shared/UrlUtils.ts (hunk 2 above)
Comment (leon-good-life) [reply app/utils/shared/UrlUtils.ts]: [98ebca5](https://github.com/cloudify-cosmo/cloudify-stage/pull/2127/commits/98ebca529353bf540393f06ac0b64fcc5a736eaa)

File: backend/handler/services/common.ts (hunk 4 above)
Comment (leon-good-life) [reply backend/handler/services/common.ts]: I addressed it: e4247ae, 630318b, 191c0e2

File: test/cypress/support/resource_commons.ts (hunk 3 above)
Comment (leon-good-life) [reply test/cypress/support/resource_commons.ts]: I addressed it: [b66d2f7](https://github.com/cloudify-cosmo/cloudify-stage/pull/2127/commits/b66d2f7df9273ab0615d8d7d73ce8221c7a61a60)

**Comment by leon-good-life** [comment]
I started system tests: https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/3431/

**Comment by leon-good-life** [comment]
All tests passed except `widgets/filters_spec.ts`. I will check  `widgets/filters_spec.ts`.

**Comment by leon-good-life** [comment]
Locally the test passes:

Screen Shot 2022-08-10 at 21 37 35

I will try to run the test again on Jenkins:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3436/pipeline

**Comment by leon-good-life** [comment]
all system tests passed

**Comment by leon-good-life** [comment]
The test passed on Jenkins

File: backend/sharedUtils.ts [hunk 5]
```diff
@@ -8,3 +11,13 @@ export function isYamlFile(filename: string) {
     const lowercaseFilename = filename.toLowerCase();
     return lowercaseFilename.endsWith('.yaml') || lowercaseFilename.endsWith('.yml');
 }
+
+export type QueryStringParams = Record<string, any>;
+
+export function getUrlWithQueryString(url: string, params?: QueryStringParams) {
+    if (!params || isEmpty(params)) return url;
```
Comment (kubama) [inline backend/sharedUtils.ts:18]:
```suggestion
    if (isEmpty(params)) return url;
```

Comment (leon-good-life) [reply backend/sharedUtils.ts:18]: It causes type error, maybe tslint don't understand lodash:

Screen Shot 2022-08-11 at 11 45 12

**Review by kubama — APPROVED** [review]

**Pull request merged**